// discordeno/script/helpers/integrations/getIntegrations.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getIntegrations = void 0;
const collection_js_1 = require("../../util/collection.js");
/** Returns a list of integrations for the guild. Requires the MANAGE_GUILD permission. */
async function getIntegrations(bot, guildId) {
    const results = await bot.rest.runMethod(bot.rest, "GET", bot.constants.routes.GUILD_INTEGRATIONS(guildId));
    return new collection_js_1.Collection(results.map((result) => {
        const integration = bot.transformers.integration(bot, {
            guild_id: guildId.toString(),
            id: result.id,
            name: result.name,
            type: result.type,
            enabled: result.enabled,
            syncing: result.syncing,
            role_id: result.role_id,
            enable_emoticons: result.enable_emoticons,
            expire_behavior: result.expire_behavior,
            expire_grace_period: result.expire_grace_period,
            user: result.user,
            account: result.account,
            synced_at: result.synced_at,
            subscriber_count: result.subscriber_count,
            revoked: result.revoked,
            application: result.application,
        });
        return [integration.id, integration];
    }));
}
exports.getIntegrations = getIntegrations;
