// discordeno/script/transformers/reverse/applicationCommand.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.transformApplicationCommandToDiscordApplicationCommand = void 0;
function transformApplicationCommandToDiscordApplicationCommand(bot, payload) {
    return {
        id: bot.transformers.reverse.snowflake(payload.id),
        application_id: bot.transformers.reverse.snowflake(payload.applicationId),
        guild_id: payload.guildId ? bot.transformers.reverse.snowflake(payload.guildId) : undefined,
        name: payload.name,
        name_localizations: payload.nameLocalizations,
        description: payload.description,
        description_localizations: payload.descriptionLocalizations,
        default_member_permissions: payload.defaultMemberPermissions
            ? bot.transformers.reverse.snowflake(payload.defaultMemberPermissions)
            : null,
        dm_permission: payload.dmPermission,
        type: payload.type,
        version: payload.version,
        options: payload.options?.map((option) => bot.transformers.applicationCommandOption(bot, option)),
    };
}
exports.transformApplicationCommandToDiscordApplicationCommand = transformApplicationCommandToDiscordApplicationCommand;
