// discordeno/script/helpers/interactions/commands/createApplicationCommand.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isContextApplicationCommand = exports.makeOptionsForCommand = exports.createApplicationCommand = void 0;
const mod_js_1 = require("../../../mod.js");
/**
 * There are two kinds of Application Commands: global commands and guild commands. Global commands are available for every guild that adds your app; guild commands are specific to the guild you specify when making them. Command names are unique per application within each scope (global and guild). That means:
 *
 * - Your app **cannot** have two global commands with the same name
 * - Your app **cannot** have two guild commands within the same name **on the same guild**
 * - Your app **can** have a global and guild command with the same name
 * - Multiple apps **can** have commands with the same names
 *
 * Global commands are cached for **1 hour**. That means that new global commands will fan out slowly across all guilds, and will be guaranteed to be updated in an hour.
 * Guild commands update **instantly**. We recommend you use guild commands for quick testing, and global commands when they're ready for public use.
 */
async function createApplicationCommand(bot, command, guildId) {
    const result = await bot.rest.runMethod(bot.rest, "POST", guildId
        ? bot.constants.routes.COMMANDS_GUILD(bot.applicationId, guildId)
        : bot.constants.routes.COMMANDS(bot.applicationId), isContextApplicationCommand(command)
        ? {
            name: command.name,
            name_localizations: command.nameLocalizations,
            type: command.type,
            default_member_permissions: command.defaultMemberPermissions
                ? bot.utils.calculateBits(command.defaultMemberPermissions)
                : undefined,
            dm_permission: command.dmPermission,
        }
        : {
            name: command.name,
            name_localizations: command.nameLocalizations,
            description: command.description,
            description_localizations: command.descriptionLocalizations,
            type: command.type,
            options: command.options ? makeOptionsForCommand(command.options) : undefined,
            default_member_permissions: command.defaultMemberPermissions
                ? bot.utils.calculateBits(command.defaultMemberPermissions)
                : undefined,
            dm_permission: command.dmPermission,
        });
    return bot.transformers.applicationCommand(bot, result);
}
exports.createApplicationCommand = createApplicationCommand;
function makeOptionsForCommand(options) {
    return options.map((option) => ({
        type: option.type,
        name: option.name,
        name_localizations: option.nameLocalizations,
        description: option.description,
        description_localizations: option.descriptionLocalizations,
        required: option.required,
        choices: option.choices?.map((choice) => ({
            name: choice.name,
            name_localizations: choice.nameLocalizations,
            value: choice.value,
        })),
        options: option.options ? makeOptionsForCommand(option.options) : undefined,
        channel_types: option.channelTypes,
        autocomplete: option.autocomplete,
        min_value: option.minValue,
        max_value: option.maxValue,
        min_length: option.minLength,
        max_length: option.maxLength,
    }));
}
exports.makeOptionsForCommand = makeOptionsForCommand;
function isContextApplicationCommand(cmd) {
    return cmd.type === mod_js_1.ApplicationCommandTypes.Message || cmd.type === mod_js_1.ApplicationCommandTypes.User;
}
exports.isContextApplicationCommand = isContextApplicationCommand;
