// discordeno/script/helpers/interactions/commands/upsertApplicationCommand.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.upsertApplicationCommand = void 0;
const createApplicationCommand_js_1 = require("./createApplicationCommand.js");
/**
 * Edit an existing application command. If this command did not exist, it will create it.
 */
async function upsertApplicationCommand(bot, commandId, command, guildId) {
    const result = await bot.rest.runMethod(bot.rest, "PATCH", guildId
        ? bot.constants.routes.COMMANDS_GUILD_ID(bot.applicationId, guildId, commandId)
        : bot.constants.routes.COMMANDS_ID(bot.applicationId, commandId), (0, createApplicationCommand_js_1.isContextApplicationCommand)(command)
        ? {
            name: command.name,
            name_localizations: command.nameLocalizations,
            type: command.type,
            default_member_permissions: command.defaultMemberPermissions
                ? bot.utils.calculateBits(command.defaultMemberPermissions)
                : undefined,
            dm_permission: command.dmPermission,
        }
        : {
            name: command.name,
            name_localizations: command.nameLocalizations,
            description: command.description,
            description_localizations: command.descriptionLocalizations,
            type: command.type,
            options: command.options ? (0, createApplicationCommand_js_1.makeOptionsForCommand)(command.options) : undefined,
            default_member_permissions: command.defaultMemberPermissions
                ? bot.utils.calculateBits(command.defaultMemberPermissions)
                : undefined,
            dm_permission: command.dmPermission,
        });
    return bot.transformers.applicationCommand(bot, result);
}
exports.upsertApplicationCommand = upsertApplicationCommand;
