// discordeno/script/helpers/guilds/automod/createAutomodRule.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createAutomodRule = void 0;
/** Get a rule currently configured for guild. */
async function createAutomodRule(bot, guildId, options) {
    const result = await bot.rest.runMethod(bot.rest, "POST", bot.constants.routes.AUTOMOD_RULES(guildId), {
        name: options.name,
        event_type: options.eventType,
        trigger_type: options.triggerType,
        trigger_metadata: {
            keyword_filter: options.triggerMetadata.keywordFilter,
            presets: options.triggerMetadata.presets,
            allow_list: options.triggerMetadata.allowList,
        },
        actions: options.actions.map((action) => ({
            type: action.type,
            metadata: action.metadata
                ? {
                    channel_id: action.metadata.channelId?.toString(),
                    duration_seconds: action.metadata.durationSeconds,
                }
                : undefined,
        })),
        enabled: options.enabled ?? true,
        exempt_roles: options.exemptRoles?.map((id) => id.toString()),
        exempt_channels: options.exemptChannels?.map((id) => id.toString()),
        reason: options.reason,
    });
    return bot.transformers.automodRule(bot, result);
}
exports.createAutomodRule = createAutomodRule;
