// discordeno/script/transformers/presence.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.transformPresence = void 0;
const shared_js_1 = require("../types/shared.js");
const user_js_1 = require("./toggles/user.js");
function transformPresence(bot, payload) {
    const presence = {
        user: {
            id: bot.transformers.snowflake(payload.user.id || ""),
            username: payload.user.username ?? undefined,
            discriminator: payload.user.discriminator ?? undefined,
            avatar: payload.user.avatar ? bot.utils.iconHashToBigInt(payload.user.avatar) : undefined,
            locale: payload.user.locale ?? undefined,
            email: payload.user.email ?? undefined,
            flags: payload.user.flags ?? undefined,
            premiumType: payload.user.premium_type ?? undefined,
            publicFlags: payload.user.public_flags ?? undefined,
            toggles: new user_js_1.UserToggles(payload.user),
        },
        guildId: bot.transformers.snowflake(payload.guild_id),
        status: shared_js_1.PresenceStatus[payload.status],
        activities: payload.activities.map((activity) => bot.transformers.activity(bot, activity)),
        desktop: payload.client_status.desktop,
        mobile: payload.client_status.mobile,
        web: payload.client_status.web,
    };
    return presence;
}
exports.transformPresence = transformPresence;
